### ISSUE TITLE
remote.getGlobal() performance

### ISSUE THREAD

**Issue by shrewdlogarithm** [atom/atom-shell#1258]
I have a few apps which pass data between BrowserWindows via remote.getGlobal() 

In reply to another issue, it was suggested that using JSON.parse(JSON.stringify(remote.getGlobal("thing"))) would force all of "thing" to be sent at-once - and certainly this seems to improve performance over accessing "thing" at-random (as I access all the content anyway)

The overall performance of this is still VERY slow tho - I have 2 globals I send regularly, one is an array, the other an object.  They have around 1000 entries/properties - are 100-150K of data in total - and it takes around 6-8 seconds to send them - that's a LONG time!

I assume this just isn't a good way of throwing data around inside an app?   I need the data back in main.js so I have to pass-it-back there - however I could always store the data in localStorage as-well, which would allow the BrowserWindows to access it without resorting to IPC?

Are there any better ways of doing this/any other tricks to speed-up this sort of stuff??

**Comment by shrewdlogarithm** [comment]
Just to update this - yes, using localStorage to pass-around data is FAR FAR quicker than remote.getGlobal

It's slightly annoying having to do both (as main.js has no access to localStorage?) tho?

**Comment by zcbenz** [comment]
You should probably just write your data into a file, IPC with big data is always slow.

**Comment by MrBlenny** [comment]
The fastest way (from what I can tell) is to Stringify the Object in the main thread:
``` javascript
global.myStringifiedLargeObject= JSON.stringify(myLargeObject);
```

You can then access this in the renderer using:
``` javascript
import { remote } from 'electron'

const myLargeObject = JSON.parse(remote.getGlobal('myStringifiedLargeObject'))
```

### Speed Comparison:
I've tried all three methods on an approx 200kB object.
* Method 1: JSON.parse(JSON.stringify()) with getGlobal - Approx 800ms
* Method 2: Writing the JSON to disk and then reading it - Read takes 400ms (with a good SSD)
* Method 3: As described above - This takes just under 10ms

**Comment by 1-14x0r** [comment]
great info. thank you.. i was really curious why this technique is fastest. After poking around in the electron source code i came across this:

[electron/lib/renderer/api/remote.js](https://github.com/electron/electron/blob/8736a41cfbc9aec4ba2ecde6788a7afccb426e40/lib/renderer/api/remote.js)
```
// Get a global object in browser.
exports.getGlobal = (name) => {
  const command = 'ELECTRON_BROWSER_GLOBAL'
  const meta = ipcRenderer.sendSync(command, name)
  return metaToValue(meta)
}
```

and 

[electron/lib/renderer/api/ipc-renderer.js](https://github.com/electron/electron/blob/7ef69a5af595c9c0eef27bdf9446103776917aec/lib/renderer/api/ipc-renderer.js)

```
ipcRenderer.sendSync = function (...args) {
  return JSON.parse(binding.sendSync('ipc-message-sync', args))
}
```

```
ipcRenderer.send = function (...args) {
  return binding.send('ipc-message', args)
}
```

first i thought the metaToValue() was a bottle neck.. but after experimenting with it, it just is a simple type map look up (in other words its a hacky way in js to cast an object to a specific type) This has almost not over head.

what i find interesting is that anytime you call remote. you basically are calling a synchronized IPC call that is blocking.. sorta makes sense i guess, as most threads should be blocking for concurrency.

JSON.parse is an expensive call when you pass in an object; it will serialize this to a string.. so when serialize the global object ahead of time, every time you call getGlobal it makes an IPC requesting the data (which is very fast) then it deserializes the return, which is what is taking a long time. When you pass an object into parse(), it will convert it to a string buffer, then parse it back into an object. Again its one of the downsides of data serialization with javascript. We are forced to do this. You can not pass literal objects over IPC (which is a wrapper for named pipes). Objects must be marshaled over pipes.

then i decided to dig a bit deeper and figure out why we are even passing json. I guess it sounds like a good idea right? we can easily serialize objects, classes and functions. This is the magic that allows you to be able to share complex modules between the main and renderer via the remote class. Pretty neat when you think about.. However it seems expensive. 

Here is the c code for anyone interested

[/atom/renderer/api/atom_api_renderer_ipc.cc](https://github.com/electron/electron/blob/44481db1eefdf0dfb5a01e37ff4d0771b51f7552/atom/renderer/api/atom_api_renderer_ipc.cc)
```
void Send(mate::Arguments* args,
          const base::string16& channel,
          const base::ListValue& arguments) {
  RenderView* render_view = GetCurrentRenderView();
  if (render_view == nullptr)
    return;

  bool success = render_view->Send(new AtomViewHostMsg_Message(
      render_view->GetRoutingID(), channel, arguments));

  if (!success)
    args->ThrowError("Unable to send AtomViewHostMsg_Message");
}

base::string16 SendSync(mate::Arguments* args,
                        const base::string16& channel,
                        const base::ListValue& arguments) {
  base::string16 json;

  RenderView* render_view = GetCurrentRenderView();
  if (render_view == nullptr)
    return json;

  IPC::SyncMessage* message = new AtomViewHostMsg_Message_Sync(
      render_view->GetRoutingID(), channel, arguments, &json);
  bool success = render_view->Send(message);

  if (!success)
    args->ThrowError("Unable to send AtomViewHostMsg_Message_Sync");

  return json;
}
```

[/atom/browser/api/event.cc](https://github.com/electron/electron/blob/787bc8570382e98c4f204abff05b2af122e5a422/atom/browser/api/event.cc)
```
bool Event::SendReply(const base::string16& json) {
  if (message_ == nullptr || sender_ == nullptr)
    return false;

  AtomViewHostMsg_Message_Sync::WriteReplyParams(message_, json);
  bool success = sender_->Send(message_);
  message_ = nullptr;
  sender_ = nullptr;
  return success;
}
```

When you first look at it you notice that they are almost identical. However the sync message has an additional param for json. This then gets assimilated into a reply message that is returned by the method.

The other last major difference is that sendSync will return a value back; in out case for getGlobal() we are returning the value that is looked up in the `global.some_variable` namespace in the main process. Makes sense for small data structures.. Its simple and elegant, and the data serialization is baked in. However for large data arrays i would use `ipcrenderer.send()` directly. I would also recommend manually serializing your data as an `async` function. The down side to this approach would be that you will need to have two event listeners, rather then one.. You will need one in the main process that recieves the request for access a store data structure, then dispatches this data over the a subsequent event which contains the data as an arg. Then in your renderer you will listen for the return event that the main process dispatched. And yes it might be a good idea to wrap all of this up into a class for a simple API to handle this.

anyways thats what i found from my research into this topic. I am now going to experiment with this.

Lastly, one method that wasn't mentioned. is to stream the data over a socket through the local loopback. no idea what the performance and memory overhead would be for that. But i did come across another thread where someone was doing this to stream video to the renderer and worked well.

thanks again for the info above, it was very interesting topic to me.

**Comment by vasyl-shumskyi** [comment]
@zcbenz @MrBlenny were there any updates on this issue. 

I'm using in-RAM object up to 100 MB as app database and would like to write it to a storage only once - when user initiated  `app.on('before-quit', () => {`  

Is it possible? 

Writing large chunks of data frequently makes app slower. 

P/S Thanks everyone for such awesome comments!
